tests/test_tfe.py

```diff
@@ -167,26 +176,68 @@ def circuit(reused_param, other_param):
 
-        @qml.qnode(self.dev1, interface='tfe')
-        def circuit_n1s(dummy1, array, dummy2):
+        @qml.qnode(qubit_device_1_wire, interface='tfe')
+        def circuit(dummy1, array, dummy2):
             qml.RY(0.5 * array[0,1], wires=0)
             qml.RY(-0.5 * array[1,1], wires=0)
             return qml.expval(qml.PauliX(0))  # returns a scalar
 
-        @qml.qnode(self.dev1, interface='tfe')
-        def circuit_n1v(dummy1, array, dummy2):
+        grad_target = (np.array(1.), np.array([[0.5,  0.43879, 0], [0, -0.43879, 0]]), np.array(-0.4))
+        cost_target = 1.03257
+
+        args = (tfe.Variable(0.46), tfe.Variable([[2., 3., 0.3], [7., 4., 2.1]]), tfe.Variable(-0.13))
+
+        def cost(x, array, y):
+            c = tf.cast(circuit(tf.constant(0.111), array, tf.constant(4.5)), tf.float32)
+            if c.shape != tf.TensorShape([]):
+                c = c[0]  # get a scalar
+            return c +0.5*array[0,0] +x -0.4*y
+
+        with tf.GradientTape() as tape:
+            cost_res = cost(*args)
+            grad_res = np.array([i.numpy() for i in tape.gradient(cost_res, [args[0], args[2]])])
+
+        assert np.allclose(cost_res.numpy(), cost_target, atol=tol, rtol=0)
+        assert np.allclose(grad_res, np.fromiter(grad_target[::2], dtype=np.float32), atol=tol, rtol=0)
+
+    def test_qnode_array_parameters_1_vector_return(self, qubit_device_1_wire, tol):
+        """Test that QNode can take arrays as input arguments, and that they interact properly with TensorFlow"""
+
+        @qml.qnode(qubit_device_1_wire, interface='tfe')
+        def circuit(dummy1, array, dummy2):
             qml.RY(0.5 * array[0,1], wires=0)
             qml.RY(-0.5 * array[1,1], wires=0)
             return qml.expval(qml.PauliX(0)),  # note the comma, returns a 1-vector
 
-        @qml.qnode(self.dev2, interface='tfe')
-        def circuit_nn(dummy1, array, dummy2):
+        grad_target = (np.array(1.), np.array([[0.5,  0.43879, 0], [0, -0.43879, 0]]), np.array(-0.4))
+        cost_target = 1.03257
+
+        args = (tfe.Variable(0.46), tfe.Variable([[2., 3., 0.3], [7., 4., 2.1]]), tfe.Variable(-0.13))
+
+        def cost(x, array, y):
+            c = tf.cast(circuit(tf.constant(0.111), array, tf.constant(4.5)), tf.float32)
+            if c.shape != tf.TensorShape([]):
+                c = c[0]  # get a scalar
+            return c +0.5*array[0,0] +x -0.4*y
+
+        with tf.GradientTape() as tape:
+            cost_res = cost(*args)
+            grad_res = np.array([i.numpy() for i in tape.gradient(cost_res, [args[0], args[2]])])
+
+        assert np.allclose(cost_res.numpy(), cost_target, atol=tol, rtol=0)
+        assert np.allclose(grad_res, np.fromiter(grad_target[::2], dtype=np.float32), atol=tol, rtol=0)
+
+    def test_qnode_array_parameters_2_vector_return(self, qubit_device_2_wires, tol):
+        """Test that QNode can take arrays as input arguments, and that they interact properly with TensorFlow"""
```

Comment:
docstrings all the same, perhaps adjust them each slightly

Reply:
Will do.